src/modules/tlsauth/tlsauth.c

```diff
@@ -0,0 +1,231 @@
+/* TLS Authentication module -- Handle automatic user authentication
+ * based on TLS client side certificate attributes.
+ *
+ * -----------------------------------------------------------------------------
+ *
+ * Copyright (c) 2020, Redis Labs
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ *   * Redistributions of source code must retain the above copyright notice,
+ *     this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *   * Neither the name of Redis nor the names of its contributors may be used
+ *     to endorse or promote products derived from this software without
+ *     specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/* To see this module in action, follow these steps:
+ *
+ */
+
+#define REDISMODULE_EXPERIMENTAL_API
+#include <string.h>
+#include <sys/types.h>
+
+#include <openssl/x509.h>
+#include <openssl/bio.h>
+#include <openssl/pem.h>
+
+#include "redismodule.h"
+
+struct AttrCheck {
+    int nid;
+    const char *req_value;
+    struct AttrCheck *next;
+};
+
+static int configUserAttr = NID_commonName;
+static struct AttrCheck *configAttrChecks = NULL;
+
+static struct AttrCheck *parseAttrCheck(RedisModuleCtx *ctx, char *attr_check_str)
```

Comment:
It's probably worth adding a check that an attribute check isn't provided twice, since that would not-obviously prevent all authentication.

Reply:
It shouldn't prevent authentication, just be inefficient.